apps/cyberstorm-nextjs/utils/auth.ts

```diff
@@ -0,0 +1,145 @@
+import { NextResponse } from "next/server";
+import type { NextRequest } from "next/server";
+import hkdf from "@panva/hkdf";
+import { EncryptJWT, jwtDecrypt } from "jose";
+import { v4 as uuid } from "uuid";
+
+interface UserStatusType {
+  username: string;
+  capabilities: string[];
+  subscription: { expires: string | null };
+  rated_packages: string[];
+  teams: string[];
+}
+
+function isUserStatusType(data: any): boolean {
+  if (!data.username) return false;
+  if (!data.capabilities) return false;
+  if (
+    typeof data.subscription.expires !== "string" &&
+    data.subscription.expires !== null
+  )
+    return false;
+  if (!data.subscription) return false;
+  if (!data.rated_packages) return false;
+  if (!data.teams) return false;
+  return true;
+}
+
+const now = () => (Date.now() / 1000) | 0;
+
+async function getDerivedEncryptionKey(secret: string | Buffer) {
+  return await hkdf(
+    "sha256",
+    secret,
+    "",
+    "Cyberstorm Generated Encryption Key",
+    32
+  );
+}
+
+async function encode(token: { username: string; expires: number }) {
+  if (process.env.CYBERSTORM_AUTH_SECRET) {
+    const encryptionSecret = await getDerivedEncryptionKey(
+      process.env.CYBERSTORM_AUTH_SECRET
+    );
+    return await new EncryptJWT(token)
+      .setProtectedHeader({ alg: "dir", enc: "A256GCM" })
+      .setIssuedAt()
+      .setExpirationTime(Math.min(now() + 900, token.expires))
+      .setJti(uuid())
+      .encrypt(encryptionSecret);
+  } else {
+    throw new Error("CYBERSTORM_AUTH_SECRET is missing");
+  }
+}
+
+export async function decode(
+  token: string
+): Promise<Record<string, unknown> | null> {
+  if (process.env.CYBERSTORM_AUTH_SECRET) {
+    const encryptionSecret = await getDerivedEncryptionKey(
+      process.env.CYBERSTORM_AUTH_SECRET
+    );
+    const { payload } = await jwtDecrypt(token, encryptionSecret, {
+      clockTolerance: 15,
+    });
+    return payload;
+  } else {
+    throw new Error("CYBERSTORM_AUTH_SECRET is missing");
+  }
+}
+
+async function fetchSubStatus(sessionid: string | undefined): Promise<{
+  username: string | null;
+  subActive: boolean;
+  expires: number;
+}> {
+  if (sessionid) {
+    const res = await fetch(
+      `https://thunderstore.io/api/experimental/current-user/`,
+      {
+        headers: {
+          authorization: `Session ${sessionid}`,
+          "Content-Type": "application/json",
+        },
+      }
+    );
+    if (res.status === 200) {
+      const data = await res.json();
+      if (isUserStatusType(data)) {
+        const currentUserStatus = data as UserStatusType;
+        if (currentUserStatus.subscription.expires) {
+          const expiryTimeEpoch = Date.parse(
```

Comment:
Date.parse might return NaN I think. What happens if it's compared against Date.now?

Reply:
I added a `!isNan` check